pkg/discovery/worker/controller_metrics_collector.go

```diff
@@ -0,0 +1,171 @@
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/metrics"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/repository"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/task"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/util"
+	api "k8s.io/api/core/v1"
+)
+
+// Collect K8s controller info from the pods by the given discovery worker and convert to KubeController objects.
+type ControllerMetricsCollector struct {
+	podList     []*api.Pod
+	cluster     *repository.ClusterSummary
+	metricsSink *metrics.EntityMetricSink
+	workerId    string
+}
+
+func NewControllerMetricsCollector(discoveryWorker *k8sDiscoveryWorker, currTask *task.Task) *ControllerMetricsCollector {
+	metricsCollector := &ControllerMetricsCollector{
+		podList:     currTask.PodList(),
+		cluster:     currTask.Cluster(),
+		metricsSink: discoveryWorker.sink,
+		workerId:    discoveryWorker.id,
+	}
+	return metricsCollector
+}
+
+// Collect allocation resource metrics for K8s controllers where usage values are aggregated from pods and capacity values
+// are from namespace quota capacity.
+func (collector *ControllerMetricsCollector) CollectControllerMetrics() ([]*repository.KubeController, error) {
+	if collector.cluster == nil {
+		return nil, fmt.Errorf("error collecting K8s controller metrics, because cluster summary object is null for discovery worker %s",
+			collector.workerId)
+	}
+	kubeNamespaceMap := collector.cluster.NamespaceMap
+	// Map from controller UID to the corresponding kubeController
+	kubeControllersMap := make(map[string]*repository.KubeController)
+	var kubeControllerList []*repository.KubeController
+	for _, pod := range collector.podList {
+		if pod.OwnerReferences == nil {
+			// If pod has no OwnerReferences, it is a bare pod without controller. Skip this.
+			glog.V(4).Infof("Skip creating KubeController for bare Pod %s", util.PodKeyFunc(pod))
+			continue
+		}
+		controllerType, controllerName, controllerUID, err := collector.getKubeControllerInfo(metrics.PodType, util.PodKeyFunc(pod))
+		if err != nil {
+			glog.Errorf("Error getting controller info from Pod %s: %v", util.PodKeyFunc(pod), err)
+			continue
+		}
+		kubeController, exists := kubeControllersMap[controllerUID]
+		if !exists {
+			namespace := pod.Namespace
+			// Create default KubeController entity
+			kubeController = repository.NewKubeController(collector.cluster.Name, namespace, controllerName, controllerType,
+				controllerUID)
+			kubeNamespace, namespaceExists := kubeNamespaceMap[namespace]
+			if !namespaceExists {
+				glog.Errorf("Namespace %s does not exist in cluster %s", namespace, collector.cluster.Name)
+				continue
+			}
+			// Add quota resources to KubeController entity with capacity as namespace quota capacity
+			for _, resourceType := range metrics.QuotaResources {
+				namespaceQuotaResource, err := kubeNamespace.GetAllocationResource(resourceType)
+				resourceCapacity := repository.DEFAULT_METRIC_CAPACITY_VALUE
```

Comment:
I think setting the quota value to unlimited for namespace DTOs (if no quota exists) and accordingly the same for the workloadController is erroneous. I think it should be set to the clusters total usable resource capacity; atleast for the requests (as limits can overshoot the cluster/nodes capacity). The clusters capacity equal the value summed up for worker/schedulable nodes.
Let me know your thoughts!

Reply:
This is exactly what @dblinn and I initially thought. It makes sense that if namespace quota is not configured, we can set the capacity as the total usable resource capacity of the whole cluster for requests, while the capacity of limits quota commodities should be much larger than cluster capacity to allow users to overcommit resource limits.

Then we considered minimizing the impact on Market analysis result if quota is not configured. In analysis we want to only resize down requests not resize up. So setting the capacity to infinity for not only limits but also requests will have the least effect on analysis result so that resource requests won't be resized from the missing namespace quota.

Reply:
Makes sense, thanks for the explaination!